examples/demo-unused-variable-warner.rkt

```diff
@@ -0,0 +1,110 @@
+#lang racket
+(require "../main.rkt"
+         (for-syntax nanopass/base
+                     syntax/name
+                     syntax/id-set)
+         racket/base)
+
+(provide (except-out (all-from-out racket/base) #%module-begin))
+
+(begin-for-syntax
+  (define (demo prog)
+    (define defined-variables (make-hasheqv))
+    (define used-variables (make-hasheqv))
+    
+    (define (mark-defined! idt phase)
+      (unless (hash-ref defined-variables phase #f)
+        (hash-set! defined-variables phase (mutable-free-id-set #:phase phase)))
+      (free-id-set-add! (hash-ref defined-variables phase)
+                        idt)
+      )
+    (define (mark-used! idt phase)
+      (unless (hash-ref used-variables phase #f)
+        (hash-set! used-variables phase (mutable-free-id-set #:phase phase)))
+      (free-id-set-add! (hash-ref used-variables phase)
+                        idt))
+    
+    (define (collect-ids! ids phase)
+      (syntax-case ids ()
+        [() (void)]
+        [(a . b) (mark-defined! #'a phase)
+                 (collect-ids! #'b phase)]
+        [(a) (mark-defined! #'a phase)]))
+    
+    (define-pass mark-unused-variable : (FullyExpandedProgram ModuleBeginForm) (prog) -> (FullyExpandedProgram ModuleBeginForm) ()
+      (ModuleBeginForm : ModuleBeginForm (prog phase) -> ModuleBeginForm ()
+                       [(#%plain-module-begin ,s ,[ml*] ... ,[ml])
+                        prog])
+      (Expr : Expr (prog phase) -> Expr ()
+            [,x 
+             (mark-used! x phase)
+             prog]
+            [(#%top ,s ,x)
+             (mark-used! x phase)
+             prog]
+            [(#%variable-reference ,s ,x)
+             (mark-used! x phase)
+             prog]
+            [(#%variable-reference-top ,s ,x)
+             (mark-used! x phase)
+             prog]
+            [(#%plain-lambda ,s ,fml ,[body*] ... ,[body])
+             (collect-ids! fml phase)
+             prog]
+            [(case-lambda ,s [,fml ,[body*] ... ,[body]] ...)
+             (for ([f fml])
+               (collect-ids! f phase))
+             prog]
+            [(let-values ,s ([(,x* ...) ,[e]] ...) ,[body*] ... ,[body])
+             (for ([x x*])
+               (collect-ids! x phase))
+             prog]
+            [(letrec-values ,s ([(,x* ...) ,[e]] ...) ,[body*] ... ,[body])
+             (for ([x x*])
+               (collect-ids! x phase))
+             prog]
+            )
+      (GeneralTopLevelForm : GeneralTopLevelForm (prog phase) -> GeneralTopLevelForm ()
+                           [(define-values ,s (,x ...) ,[e])
+                            (collect-ids! x phase)
```

Comment:
I think we should ignore module-level definitions since they could be provided or accessed via namespaces.